src/runtime/program/vote/execute.zig

```diff
@@ -409,6 +429,249 @@ fn executeAuthorizeChecked(
+            ) catch {
+                break :blk true;
+            };
+            const vote_state = try versioned_state.convertToCurrent(allocator);
+            maybe_vote_state = vote_state;
+            break :blk vote_state.isCommissionIncrease(commission);
+        } else {
+            break :blk true;
+        }
+    };
+
+    if (enforce_commission_update_rule and ic.tc.feature_set.isActive(
+        feature_set.COMMISSION_UPDATES_ONLY_ALLOWED_IN_FIRST_HALF_OF_EPOCH,
+    )) {
+        if (!isCommissionUpdateAllowed(clock.slot, &epoch_schedule)) {
+            // Clean up before returning, if we have a vote_state already.
+            if (maybe_vote_state) |*vote_state| {
+                vote_state.deinit();
+            }
+            ic.tc.custom_error = @intFromEnum(VoteError.commission_update_too_late);
+            return InstructionError.Custom;
+        }
+    }
+
+    var vote_state = blk: {
+        if (maybe_vote_state) |vote_state| {
+            break :blk vote_state;
+        } else {
+            const versioned_state = try vote_account.deserializeFromAccountData(
+                allocator,
+                VoteStateVersions,
+            );
+
+            break :blk versioned_state.convertToCurrent(allocator) catch {
+                return InstructionError.InvalidAccountData;
+            };
+        }
+    };
+    defer vote_state.deinit();
+
+    // Current authorized withdrawer must sign transaction.
+    if (!ic.info.isPubkeySigner(vote_state.authorized_withdrawer)) {
+        return InstructionError.MissingRequiredSignature;
+    }
+
+    vote_state.commission = commission;
+    try vote_account.serializeIntoAccountData(VoteStateVersions{ .current = vote_state });
+}
+
+/// [agave] https://github.com/anza-xyz/agave/blob/a7092a20bb2f5d16375bdc531b71d2a164b43b93/programs/vote/src/vote_state/mod.rs#L798
+///
+/// Given the current slot and epoch schedule, determine if a commission change
+/// is allowed
+pub fn isCommissionUpdateAllowed(slot: u64, epoch_schedule: *const EpochSchedule) bool {
+    // Always allowed during warmup epochs
+    const maybe_relative_slot: ?u64 = std.math.rem(
+        u64,
+        (slot -| epoch_schedule.first_normal_slot),
+        epoch_schedule.slots_per_epoch,
+    ) catch null;
+
+    if (maybe_relative_slot) |relative_slot| {
+        return (relative_slot *| 2 <= epoch_schedule.slots_per_epoch);
+    } else {
+        return true;
+    }
+}
+
+/// [agave] https://github.com/anza-xyz/agave/blob/e363f52b5bb4bfb131c647d4dbd6043d23575c78/programs/vote/src/vote_processor.rs#L222-L227
+fn executeWithdraw(
+    allocator: std.mem.Allocator,
+    ic: *InstructionContext,
+    vote_account: *BorrowedAccount,
+    lamports: u64,
+) (error{OutOfMemory} || InstructionError)!void {
+    try ic.info.checkNumberOfAccounts(2);
+    const rent = try ic.getSysvar(Rent);
+    const clock = try ic.getSysvar(Clock);
+
+    // TODO agave drops the vote_account here. Why?
```

Comment:
Not sure why the accounts are manually dropped and then retrieved in Agave, but will need to update this to match the drop semantics.

Reply:
[1ffa48c](https://github.com/Syndica/sig/pull/616/commits/1ffa48cb140d4c2c37a54d68e0bb393aec2fc558)